car-bridge/src/main.rs

```diff
@@ -29,58 +41,198 @@ async fn main() -> Result<(), Box<dyn Error>> {
+                        debug!("Shutting down the publisher loop.");
+                        break;
+                    }
+                }
+            }
+        })
+    };
 
     loop {
         select! {
-            message = message_stream.next() => {
+            message = messages.next() => {
                 let Some(message) = message else {
                     break;
                 };
-                if let Some(message) = message {
-                    info!("(R) {message:?}");
-                }
-                else {
-                    info!("Connection temporarily lost.");
 
-                    while let Err(err) = client.reconnect().await {
-                        info!("Trying to reconnect: {}.", err);
-                        sleep(Duration::from_secs(5)).await;
-                    }
-                }
+                let mut chariott = chariott.clone();
+                let response_sender = response_sender.clone();
+
+                spawn(async move {
+                    handle_message(&mut chariott, response_sender, message).await;
+                });
             }
             _ = cancellation_token.cancelled() => {
+                debug!("Shutting down the subscriber loop.");
                 break;
             }
         }
     }
 
-    info!("Disconnecting the client.");
-    client.disconnect(None).await?;
+    publish_handle.await?;
 
     Ok(())
 }
+
+async fn handle_message(
+    chariott: &mut impl ChariottCommunication,
+    response_sender: Sender<(String, MessageBuilder)>,
+    message: MqttMessage,
+) {
+    async fn get_response(
+        chariott: &mut impl ChariottCommunication,
+        message: &MqttMessage,
+    ) -> Result<Response, Box<dyn std::error::Error>> {
+        let fulfill_request: FulfillRequest = Message::decode(message.payload())?;
+
+        let intent_enum = fulfill_request
+            .intent
+            .and_then(|intent| intent.intent)
+            .ok_or_else(|| Error::new("Message does not contain an intent."))?;
+
+        let response = match intent_enum {
+            IntentEnum::Discover(_) => Err(Error::new("Discover is not supported.")),
+            IntentEnum::Subscribe(_) => todo!(),
+            _ => chariott
+                .fulfill(fulfill_request.namespace, intent_enum)
+                .await
+                .map(|r| r.into_inner()),
+        }?;
+
+        let mut payload = vec![];
+        response.encode(&mut payload)?;
+
+        Ok(Response {
+            payload,
+            content_type: "application/x-proto+chariott.common.v1.FulfillResponse",
+            is_error: false,
+        })
+    }
+
+    let correlation_information = match message.get_correlation_information() {
+        Ok(cm) => cm,
+        Err(error) => {
+            debug!("Error when getting correlation information from message: '{error:?}'.");
+            return;
+        }
+    };
+
+    let response = match get_response(chariott, &message).await {
```

Comment:
I think you can get rid of the local function `get_response` by just inlining it as a fallible async block:

```suggestion
    let correlation_information = match message.get_correlation_information() {
        Ok(cm) => cm,
        Err(error) => {
            debug!("Error when getting correlation information from message: '{error:?}'.");
            return;
        }
    };

    let response: Result<_, Box<dyn std::error::Error + Send + Sync>> = async {
        let fulfill_request: FulfillRequest = Message::decode(message.payload())?;

        let intent_enum = fulfill_request
            .intent
            .and_then(|intent| intent.intent)
            .ok_or_else(|| Error::new("Message does not contain an intent."))?;

        let response = match intent_enum {
            IntentEnum::Discover(_) => Err(Error::new("Discover is not supported.")),
            IntentEnum::Subscribe(_) => todo!(),
            _ => chariott
                .fulfill(fulfill_request.namespace, intent_enum)
                .await
                .map(|r| r.into_inner()),
        }?;

        let mut payload = vec![];
        response.encode(&mut payload)?;

        Ok(Response {
            payload,
            content_type: "application/x-proto+chariott.common.v1.FulfillResponse",
            is_error: false,
        })
    }
    .await;

    let response = match response {
```

It keeps the flow sequential: get correlation info or exit, get response or exit, etc.